### PR TITLE
Implemented  UI-Register-Form with email/password input validation

### PR DESCRIPTION
**Pull request by ahuzmeza**

### CODE REVIEW

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+import UIKit
+
+class RegisterViewController: UIViewController {
+    // outlets
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var personalIDTextfield: UITextField!
+    @IBOutlet weak var passwordTextfield: UITextField!
+    @IBOutlet weak var studentIDTextfield: UITextField!
+    @IBOutlet weak var registerButton: UIButton!
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:16]: add a new line between sections, not just here

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    // actions
+    @IBAction func emailCHanged(_ sender: Any) {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:18]: typo: emailChanged

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        validateIfFormFilled()
+    }
+    @IBAction func passwordChanged(_ sender: Any) {
+        validateIfFormFilled()
+    }
+    @IBAction func personal_IDChanged(_ sender: Any) {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:24]: avoid "_", use personaldChanged

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        validateIfFormFilled()
+    }
+    @IBAction func student_IDChanged(_ sender: Any) {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:27]: avoid "_"

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        validateIfFormFilled()
+    }
+    @IBAction func submitAction(_ sender: Any) {
+        handleFormValidation()
+    }
+    // input validation error strings
+    var emailError: String = ""
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:34]: use private

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    var passwordError: String = ""
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:35]: use private

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            }
+        }
+        // Add style to button
+        addStyleToButton(button: registerButton)
+        // Set initial state of 'registerButton' to disabled
+        registerButton.isEnabled = false
+    }
+    
+    func handleFormValidation() {
+        let email = emailTextField.text!
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:55]: avoid using "!", instead you can write emailTextField.text ?? ""

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        // enable/disable 'register' button
+        if foundEmpty {
+            registerButton.isEnabled = false
+        } else {
+            registerButton.isEnabled = true
+        }
+    }
+    
+    func isEmptyTextfield(_ textField: UITextField) -> Bool {
+        if textField.text!.isEmpty {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:117]: avoid using "!", check for nil or empty

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            return true
+        }
+        return false
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:120]: you can return just the result of: textField.text!.isEmpty

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        addStyleToButton(button: registerButton)
+        // Set initial state of 'registerButton' to disabled
+        registerButton.isEnabled = false
+    }
+    
+    func handleFormValidation() {
+        let email = emailTextField.text!
+        let password =  passwordTextfield.text!
+        
+        emailError = validateEmail(email) ?? ""
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:58]: use if emailError then showAlert with email
else if passwordError then showAlert with password
else continue to register

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        for textField in self.view.subviews {
+            if type(of: textField) == UITextField.self {
+                if isEmptyTextfield(textField as! UITextField) {
+                    foundEmpty = true
+                    break
+                }
+            }
+        }
+        // enable/disable 'register' button
+        if foundEmpty {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:109]: use registerButton.isEnabled = !foundEmpty

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    }
+    
+    func isEmptyTextfield(_ textField: UITextField) -> Bool {
+        if textField.text!.isEmpty {
+            return true
+        }
+        return false
+    }
+    
+    func alertError(_ errors: [String]) {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:123]: show alert for only one string

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+import UIKit
+
+class RegisterViewController: UIViewController {
+    // outlets
+    @IBOutlet weak var emailTextField: UITextField!
+    @IBOutlet weak var personalIDTextfield: UITextField!
+    @IBOutlet weak var passwordTextfield: UITextField!
+    @IBOutlet weak var studentIDTextfield: UITextField!
+    @IBOutlet weak var registerButton: UIButton!
+    // actions
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:17]: put actions after view did load, they are methods

**Review by iOSinternshipen — CHANGES_REQUESTED** [review]

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift (hunk 1 above)
Comment (iOSinternshipen) [reply Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:16]: also private outlets / actions

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift (hunk 8 above)
Comment (ahuzmeza) [reply Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:117]: Can't figure it out. Would appreciate example

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 14]
```diff
@@ -106,26 +111,18 @@ class RegisterViewController: UIViewController {
...
     }
     
     func isEmptyTextfield(_ textField: UITextField) -> Bool {
-        if textField.text!.isEmpty {
-            return true
+        if textField.text != nil {
+            return textField.text!.isEmpty
         }
-        return false
+        return true
```
Comment (IuliaRAnisoi) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:121]: I believe you can simply write "return textField.text?.isEmpty ?? true"

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 15]
```diff
@@ -106,26 +111,18 @@ class RegisterViewController: UIViewController {
...
-        var errorMessage: String = ""
-        for error in errors {
-            errorMessage += "\n\(error)"
-        }
-        let dialogMessage = UIAlertController(title: "Error", message: errorMessage, preferredStyle: .alert)
+    func alertError(_ error: String) {
+        let dialogMessage = UIAlertController(title: "Error", message: error, preferredStyle: .alert)
         // Create OK button with action handler
         let ok = UIAlertAction(title: "OK", style: .default, handler: { (action) -> Void in
         })
```
Comment (IuliaRAnisoi) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:128]: you can remove the handler parameter entirely if you don't use it

**Review by IuliaRAnisoi — APPROVED** [review]

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        // Do any additional setup after loading the view.
+        // Add style to all textfields
+        for textField in self.view.subviews {
+            if type(of: textField) == UITextField.self
+            {
+                addStyleToTextfield(textField: textField as! UITextField)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:30]: force unwrap

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    func addStyleToTextfield(textField: UITextField)
+    {
+        // border bottom
+        let bottomLine = CALayer()
+        bottomLine.frame = CGRect(
+            x: 0.0,
+            y: textField.frame.height - 1,
+            width: textField.frame.width,
+            height: 1.0)
+        bottomLine.backgroundColor = UIColor(
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:147]: add color to assets

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+    // input validation error strings
+    private var emailError: String = ""
+    private var passwordError: String = ""
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        // Do any additional setup after loading the view.
+        // Add style to all textfields
+        for textField in self.view.subviews {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:27]: 1. you can use forEach, is similar but more Swifty
2. maybe you can create a custom textfield class to be reused also in Log in screen?

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+//  Created by Alin Huzmezan on 26.07.2022.
+//
+
+import UIKit
+
+class RegisterViewController: UIViewController {
+    // outlets
+    @IBOutlet private weak var emailTextField: UITextField!
+    @IBOutlet private weak var personalIDTextfield: UITextField!
+    @IBOutlet private weak var passwordTextfield: UITextField!
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:14]: for the password textfield, please use the secure text option, so nobody can see what I'm typing :)

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+    }
+    @IBAction private func studentIDChanged(_ sender: Any) {
+        validateIfFormFilled()
+    }
+    @IBAction private func submitAction(_ sender: Any) {
+        handleFormValidation()
+    }
+    
+    func handleFormValidation() {
+        let email = emailTextField.text ?? ""
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:57]: I would use guard, and if nothing is there to show an alert

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+        let password =  passwordTextfield.text ?? ""
+        
+        emailError = validateEmail(email) ?? ""
+        passwordError = validatePassword(password) ?? ""
+        
+        if emailError != ""  {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:63]: I would use guard over if

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+                    foundEmpty = true
+                    break
+                }
+            }
+        }
+        // enable/disable 'register' button
+        registerButton.isEnabled = !foundEmpty
+    }
+    
+    func isEmptyTextfield(_ textField: UITextField) -> Bool {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:117]: this method is used?

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+        let emailPred = NSPredicate(format:"SELF MATCHES %@", emailRegEx)
+        // must be of valid type *@*.*
+        guard emailPred.evaluate(with: email) else {
+            return "Invalid email."
+        }
+        // valid email
+        return nil
+    }
+    
+    func validatePassword(_ password: String) -> String? {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:88]: I would make the method private and all the methods that are used only in this class

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift (hunk 21 above)
Comment (iOSinternshipen) [reply Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:63]: also use .isEmpty

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    @IBAction private func studentIDChanged(_ sender: Any) {
+        validateIfFormFilled()
+    }
+    @IBAction private func submitAction(_ sender: Any) {
+        handleFormValidation()
+    }
+
+    private func handleFormValidation() {
+        let emailError = validateEmail(emailTextField.text!)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:48]: dont use force unwrap, try emailTextField.text ?? ""

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+        let passwordError = validatePassword(passwordTextfield.text!)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:49]: dont use force unwrap

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+
+        guard emailError == nil else {
+            alertError(emailError!)
+            return
+        }
+        guard passwordError == nil else {
+            alertError(passwordError!)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:56]: dont use force unwrap

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+    @IBAction private func submitAction(_ sender: Any) {
+        handleFormValidation()
+    }
+
+    private func handleFormValidation() {
+        let emailError = validateEmail(emailTextField.text!)
+        let passwordError = validatePassword(passwordTextfield.text!)
+
+        guard emailError == nil else {
+            alertError(emailError!)
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:52]: dont use force unwrap, use

**Review by iOSinternshipen — CHANGES_REQUESTED** [review]

**Review by IuliaRAnisoi — APPROVED** [review]

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+    @IBOutlet private weak var emailTextField: UITextField!
+    @IBOutlet private weak var personalIDTextfield: UITextField!
+    @IBOutlet private weak var passwordTextfield: UITextField!
+    @IBOutlet private weak var studentIDTextfield: UITextField!
+    @IBOutlet private weak var registerButton: UIButton!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        view.subviews.forEach { view in
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:21]: you can move it in a method like setupAppearance or something like that

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+        let dialogMessage = UIAlertController(title: "Error", message: error, preferredStyle: .alert)
+        let okAction = UIAlertAction(title: "OK", style: .default)
+        dialogMessage.addAction(okAction)
+        self.present(dialogMessage, animated: true, completion: nil)
+    }
+
+    func addStyleToButton(button: UIButton) {
+        // shadow
+        button.layer.shadowColor = UIColor.black.cgColor
+        button.layer.shadowOffset = CGSize(
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:115]: can be on one line

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+            width: 1,
+            height: 4)
+        button.layer.shadowRadius = 2
+        button.layer.shadowOpacity = 0.2
+    }
+
+    func addStyleToTextfield(textField: UITextField) {
+        // border bottom
+        let bottomLine = CALayer()
+        bottomLine.frame = CGRect(
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:125]: same here

File: Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+            height: 1.0)
+        bottomLine.backgroundColor = UIColor(named: "TextFieldBorderColor")?.cgColor
+        textField.borderStyle = UITextField.BorderStyle.none
+        textField.layer.addSublayer(bottomLine)
+    }
+}
+
+// MARK: - for keyboard issues
+extension RegisterViewController: UITextFieldDelegate {
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
```
Comment (iOSinternshipen) [inline Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift:138]: I think you don't set the delegate anywhere

Comment (ahuzmeza) [reply Internship2022TeamSzabi/Internship2022TeamSzabi/Views/Register/RegisterViewController.swift]: Removed entirely. It was a temporary fix for keyboard that will be fixed at a later date.

**Review by iOSinternshipen — APPROVED** [review]

**Pull request merged**